File: 2025/106. Construct Binary Tree from Inorder and Postorder Traversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    Map<Integer, Integer> map;
    int index;

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        map = new HashMap<>();
        index = postorder.length - 1;
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }
        return build(0, inorder.length - 1, inorder, postorder);
    }

    private TreeNode build(int left, int right, int[] inorder, int[] postorder) {
        if (left > right) {
            return null;
        }
        int val = postorder[index--];
        TreeNode n = new TreeNode(val);
        n.right = build(map.get(val) + 1, right, inorder, postorder);
        n.left = build(left, map.get(val) - 1, inorder, postorder);
        return n;
    }
}

File: 2025/106. Construct Binary Tree from Inorder and Postorder Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        inorder_map = {item: idx for idx, item in enumerate(inorder)}
        index = len(postorder) - 1
        def build(left, right): 
            nonlocal index
            if left > right:
                return
            val = postorder[index]
            index -= 1
            node = TreeNode(val)
            node.right = build(inorder_map[val] + 1, right)
            node.left = build(left, inorder_map[val] - 1)
            return node
            
        return build(0, len(inorder) - 1)